readme.md
Welcome to Framework.

# What is Framework?

Framework is the python 3 compatible audiogame creation toolkit. It combines many things and aims to make it simple to create audiogames in Python.

Note! Some things are still undocumented!

# Requirements
* pygame
* synthizer (pip install synthizer)
* cytolk (pip install cytolk)

# Credits
This toolkit has been Created by Mason Armstrong. BGT style Sound class and sound manager created initially for Sound_Lib by Carter Temm and continued/adapted to Synthizer by Mason Armstrong. Initial sound_pool class for sound_lib created by Kianoosh and updated/finished by Mason Armstrong. Menu and map classes created by Mason Armstrong. Some code originally part of the AGK and AGK3 packages.

# building/cythonizing
For this to work, you'll need a c compiler and cython. It is not required to build framework in order to use it. Just run build.bat and it'll spit out pyd fies in a folder called fw.

# What do we have?

## Classes

### sound3d
sound class for manually manipulating sounds.
### constructor
sound3d(type, context)
type (string): direct, pannable, or 3D. Check the synthizer documentation on what each one of these does.
context: A synthizer context object.
#### Functions
sound.load("filename.extension")
Load a sound into memory.
sound.close()
close a sound from memory.
sound.play()
play a sound once.
sound.play_looped()
play a sound looped.
sound.play_wait()
play a sound and wait until it is finished.
sound.pause()
pause a sound.
sound.stop()
stop a sound

Sound Properties
active: Read only Boolean which tells if a sound is active or not.
paused: Read only Boolean which tells if a sound is paused or not.
pan: A number between -100 and 100 which specifies the pan of a sound. This only works for pannable sounds.
pitch: A number between 0 and infinity which sets the pitch of a sound. 100 is normal, above is higher and below is lower.
volume: A number which sets the volume (in DB) of a sound. 0 is loudest, below is quieter.
position: A number which sets and shows the seek position (in ms) of a sound.
length: A read only number showing the length of a sound in seconds.
playing: gets the status of a sound, playing is True.

## sound_manager
a high level class for handling sounds.
### Functions:
sound_manager.play_stationary(filename,looping)
Play a sound in the center, either one shot or looped. Returns the sound slot for later manipulation.
sound_manager.play_1d(filename, sound_x, looping)
Play a sound in a 1 Dementional setting. Returns the slot for later manipulation.
sound_manager.play_2d(filename, sound_x, sound_y, looping)
Play a sound in a 2 Dementional setting. Returns the slot for later manipulation.
sound_manager.play_3d(filename, sound_x, sound_y, sound_z, looping)
Play a sound in a 3 Dementional setting. Returns the slot for later manipulation.

sound_manager.destroy_all()
destroy all playing sounds.
sound_manager.update_sound_volume(slot, volume)
Update the sound volume.
sound_manager.update_sound_pitch(slot, pitch)
Update the sound pitch.
sound_manager.update_1d(slot, sound_x)
update a sound's position.
sound_manager.update_2d(slot, sound_x, sound_y)
update a sound's position.
sound_manager.update_3d(slot, sound_x, sound_y, sound_z)
update a sound's position.
### properties
sound_manager.x, sound_manager.y and sound_manager.z.
Update a listener's position.


## Menu
A class for creating and displaying a menu.
### menu(arguments)
All arguments are optional. Valid arguments:
music=""
set music to play in this menu.
music_volume=-20
set the initial music volume.
fade_music=False
Set if the music will fade by default or not.
fade_time=25
Set how fast the music fades.
up_and_down=True
enable the up and down arrows.
left_and_right=False
Set if left and right arrows work in the menu.
pan_sounds=False
set if menu sounds pan relative to the position you are in the menu. Doesn't work since converting to synthizer yet.
click_sound=""
set a click sound for your menu.
enter_sound=""
set a sound for when you press enter on the item.
edge_sound=""
set a sound for your menu hitting the edge.
wrap_sound=""
set a sound for a wrapping menu.
open_sound=""
set a sound for a menu opening.
escape_sound=""
set a sound for escaping out of the menu.
return_type=menu_object
set the return type, either menu_index or menu_object. If index, when you run the run option, you will be returned a menu_item which has the values text and name.
home_and_end=True
set if the menu allows home and end.
wrap=False
set if the menu wraps
select_with_enter=True
select if enter selects menu items.
select_with_space=False
select if space selects menu items.
announce_position_info=False
select if position info is announced, such as 1 of 4.

### Functions
get_item(index)
get a menu item object from an index.
add_item_tts(text_to_speak,name_of_item,item_enabled=True)
add a tts item to the menu. Disabling the item will disallow enter press on the item.
run(intro=None,tts_intro=True,starting_position=-1)
run the menu! Returned item is shown above.


## Map
This class allows you to create 1D, 2D, or 3D maps and query tiles at positions. It is new, so therefore may be missing crucial features.
### your_map=map(type)
This function accepts different perametors based on what type you select. The choices are:
type_1d
type_2d
type_3d
if you choose type=type_1d, the other perametors are:
maxx
the maximum x of this map.
tile
the default surface of this map.
if you choose type=type_2d, the other perametors are:
maxx
the maximum x of this map.
maxy
the maximum y of this map.
tile
the default surface of this map.
if you choose type=type_2d_platformer, the other perametors are:
maxx
the maximum x of this map.
maxy
the maximum y of this map.
tile
the default surface of this map.
The difference is that tiles will only be spawned the maximum x at y 0, and not all Y's.
if you choose type=type_3d, the other perametors are:
maxx
the maximum x of this map.
maxy
the maximum y of this map.
maxz
the maximum z of this map.
tile
the default surface of this map.
Example for spawning a 2D map
my_map=map(type=type_2d,tile="concrete",maxx=40,maxy=20)

### Functions
platform(arguments)
if 1d, arguments are:
minx=number, maxx=number, tile=""
if 2d, arguments are:
minx=number, maxx=number, miny=number, maxy=number, tile=""
if 3d, arguments are:
minx=number, maxx=number, miny=number, maxy=number, minz=number, maxz=number, tile=""
example
my_map.platform(minx=0,maxx=15,miny=0,maxy=4,minz=0,maxz=0,tile="rocks")

map.get_tile_at(arguments)
if 1d, arguments are
x
if 2d, arguments are
x, y
if 3d, arguments are
x, y, z
Example
tile=my_map.get_tile_at(x=5,y=3,z=0)


## Window

Taken from Lucia and adapted for Framework.

###Functions

window.initialize()
You must call this function at the top of your code. It initializes pygame.

quit()
Frees pygame.

show_window(title,size)
Shows a window on the screen. Title and size are optional.

process_events()
You must call this in your main loop, otherwise the window will become not responding.

key_pressed(key)
check if a key is pressed or not. E.G. window.key_pressed(window.K_left)

key_released(key)
check if a key is released or not. E.G. window.key_released(window.K_left)

key_down(key)
check if a key is being held down or not. E.G. window.key_down(window.K_left)

key_up(key)
check if a key is not being held down. E.G. window.key_up(window.K_left)

## timer
Timer class, taken from lucia and adapted for Framework.

### Functions
timer.restart()
Restarts the timer back to 0 MS

timer.pause()
pause a timer.

timer.resume()
resume a paused timer.

### properties

timer.elapsed
Set or get the elapsed in MS of the timer.

## Misc
dlg(text,callback=None)
show a square dialog. press arrows to repeat, enter to escape.
dlg_play(sound,callback=None,fade=False,fadespeed=30, context=None)
Play a sound and allow enter to interrupt it. You must supply a synthizer context object. Sound managers get a context that you can use, sound_manager.context.
grt(ms)
convert ms to a human readable time.
key_holding(key,delay=500,repeat=50)
see if a key is being pressed and then repeat if it is continued to be held down.
speak(text)
speak some text with the screen reader or SAPI.

changelog.md
# 2.0
* Switches to Synthizer for sound handling code. The way this works is pretty ugly right now.
* Splits code up into many files.
* Possibly going open source? IDK.
* Fixes many bugs/inconsistancies.

# 0.x-1.0
* Undocumented changes, renames to framework and removes window functions, those are now handled by Lucia.


# Older versions

# 0.28
* Now using Kianoosh's sound pool (Thanks Kianoosh!)

# 0.27
* renames window_flush to flush_window for the sake of consistency, because every other window-related func is x_window, not window_x.
* adds is_playing and is_active to sound pool.
* New functions from Ivan

# 0.26
* Some unofficial/broken functions removed.

# 0.25
* Modifications to pack file to work under Linux. Requires recompile of sounds.

# 0.24
* adds clear method to map class
* Adds map class!

# 0.23
* Adds dl_file function.

# 0.22
* Make listener_angle optional in 2d and 3d.

# 0.21
* Adds pgt_credits and pgt_version vars.
* Adds tempo and tempo_pitch variable to sound.

# 0.2
* Adds effects to sound class.

# 0.14
* Just for fun, replaced ogg sounds with Opus ones, decreasing size by 200 KB
* Now you have example bat file for EXE compilation. If you don't put the hook file in your hooks folder, expect this to fail hard.
* Now performing UPX compression. Size goes down from 400 kb to 98 kb. Boom!
* PGT is now cythonized! Source is in src folder, with tools needed to compile. Still need vs. When using, copy pgt.pyd, not py. Still need pack.py.

# 0.13
* Fade function in sound_pool's destroy_all def.
* panning in menues now defaults to false
* added the fade function to menu that optionally takes 2 parameters: end that defaults to -40, and time that defaults to self.fade_time

# 0.12
* changed the while loop in the dlg_play function to allow the function to die when the sound has finished playing rather than still waiting for you to press enter
* Adds sleep to all functions that wait for user input (Menu, dlg, etc)

# 0.11
* Menu.add_item_tts now has the disable perametor to disable menu items.

framework/menu3d.py
import pygame
import lucia
import time
from .speech import *
from .sound3d import *
from .misc import *
menu_object=1
menu_index=2
class menu_item3d(object):
	def __init__(self,text,name,is_tts,enabled):
		self.text=text
		self.name=name
		self.is_tts=is_tts
		self.enabled=enabled

class menu3d(object):
	def __init__(self, **kwargs):
		self.pool=kwargs.get('pool', None)
		self.callback=kwargs.get('callback', None)
		self.music=kwargs.get('music','')
		self.music_volume=kwargs.get('music_volume',-20)
		self.fade_music=kwargs.get('fade_music',False)
		self.fade_time=kwargs.get('fade_time',25)
		self.up_and_down=kwargs.get('up_and_down',True)
		self.left_and_right=kwargs.get('left_and_right',False)
		self.pan_sounds=kwargs.get('pan_sounds',False)
		self.click_sound=kwargs.get('click_sound','')
		self.enter_sound=kwargs.get('enter_sound','')
		self.edge_sound=kwargs.get('edge_sound','')
		self.wrap_sound=kwargs.get('wrap_sound','')
		self.open_sound=kwargs.get('open_sound','')
		self.escape_sound=kwargs.get('escape_sound','')
		self.return_type=kwargs.get('return_type',menu_object)
		self.at_edge=False
		self.position=-1
		self.items=[]
		self.home_and_end=kwargs.get('home_and_end',True)
		self.wrap=kwargs.get('wrap',False)
		self.select_with_enter=kwargs.get('select_with_enter',True)
		self.select_with_space=kwargs.get('select_with_space',False)
		self.announce_position_info=kwargs.get('announce_position_info',False)
		if self.music!="":
			self.mus=sound3d("direct",self.pool.context)
			self.mus.load(self.music)
			self.mus.volume=self.music_volume

	def get_item(self,index):
		return self.items(index-1)

	def add_item_tts(self,text,name, enabled=True):
		self.items.append(menu_item3d(text,name,True,enabled))

	def run(self,intro=None,is_intro_tts=True,starting_position=-1):
		if self.music!="":
			self.mus.play_looped()
		self.at_edge=False
		self.play(self.open_sound)
		self.position=starting_position
		if intro:
			if is_intro_tts:
				if starting_position==-1:
					speak(intro,1)
				else:
					speak(intro+". "+self.return_speak_item(),1)
			else:
				s=sound3d("direct",self.pool.context)
				s.load(intro)
				s.play()

		while 1:
			lucia.process_events()
			time.sleep(0.005)
			if self.callback:
				self.callback(self)
			if key_holding(pygame.K_PAGEUP) and self.music!="":
				if self.mus.volume<0:
					self.mus.volume+=1
			if key_holding(pygame.K_PAGEDOWN) and self.music!="":
				if self.mus.volume>-30:
					self.mus.volume-=1
			if key_holding(pygame.K_UP) and self.up_and_down==True or key_holding(pygame.K_LEFT) and self.left_and_right==True:
				self.cycle(1)
			elif key_holding(pygame.K_DOWN) and self.up_and_down==True or key_holding(pygame.K_RIGHT) and self.left_and_right==True:
				self.cycle(2)
			elif self.home_and_end:
				if lucia.key_pressed(pygame.K_HOME):
					self.at_edge=False
					self.position=0
					self.speak_item()
				elif lucia.key_pressed(pygame.K_END):
					self.at_edge=False
					self.position=len(self.items)-1
					self.speak_item()

			if (lucia.key_pressed(pygame.K_RETURN) and self.select_with_enter or lucia.key_pressed(pygame.K_SPACE) and self.select_with_space) and self.position!=-1:
				self.play(self.enter_sound)
				if self.items[self.position].enabled==True:
					self.kill_music()
					if self.return_type==menu_object:
						return self.items[self.position]
					else:
						return (self.position+1)

			if lucia.key_pressed(pygame.K_ESCAPE):
				self.play(self.escape_sound)
				self.kill_music()
				return -1

	def play(self,snd):
		if snd!="":
			if self.pan_sounds==False:
				self.pool.play_stationary(snd,False)
#			else:
#				self.pool.play_stationary_extended(snd,False, 0, convert_to_pan(len(self.items),self.position,-100,100),0,100)

	def cycle(self, dir):
		self.at_edge=False
		if dir==1:
			if self.position==-1:
				self.position=len(self.items)-1
			else:
				self.position-=1
			if self.position<0:
				if self.wrap==True:
					self.play(self.wrap_sound)
					self.position=len(self.items)-1
				else:
					self.position+=1
					self.at_edge=True
					self.play(self.edge_sound)
		if dir==2:
			if self.position==-1:
				self.position=0
			else:
				self.position+=1
			if self.position>len(self.items)-1:
				if self.wrap==True:
					self.play(self.wrap_sound)
					self.position=0
				else:
					self.position-=1
					self.at_edge=True
					self.play(self.edge_sound)
		self.speak_item()

	def speak_item(self):
		if self.at_edge==False:
			self.play(self.click_sound)
			speak(self.return_speak_item(),1)

	def return_speak_item(self):
		speaktext=self.items[self.position].text
		if self.announce_position_info==True:
			speaktext+=". "+str(self.position+1)+" of "+str(len(self.items))
		return speaktext

	def kill_music(self):
		if self.music!="":
			if self.fade_music==True:
				self.mus.fade(-40, self.fade_time)
			self.mus.stop()

	def fade(self, end=-40, time=None):
		if time==None: time=self.fade_time
		if self.music!="":
			self.mus.fade(end, time)
			self.mus.stop()

def convert_to_pan(length, index,  range1,  range2):
	range=range2-range1
	percent=0
	if length-1>0 and index>0:
		percent=index/(length-1)
	else:
		percent=0
	value=range1+range*percent
	return value
